File: code/ch13/13.6.1.Solution.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0
        dp = [0] * (len(prices)+1)
        min_price = prices[0]

        for i in range(1, len(prices)):
            min_price = min(min_price, prices[i] - dp[i - 1])
            dp[i + 1] = max(dp[i], prices[i] - min_price)
        return dp[-1]

File: code/ch13/13.6.1.Solution.java
package ch13.code;

public class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2)
            return 0;
        int[] dp = new int[prices.length + 1];
        int min = prices[0];

        for (int i = 1; i < prices.length; i++) {
            min = Math.min(min, prices[i] - dp[i - 1]);
            dp[i + 1] = Math.max(dp[i], prices[i] - min);
        }
        return dp[dp.length - 1];
    }
}